go/vt/vttls/vttls.go

```diff
@@ -30,37 +30,57 @@ import (
 // Updated list of acceptable cipher suits to address
 // Fixed upstream in https://github.com/golang/go/issues/13385
 // This removed CBC mode ciphers that are suseptiable to Lucky13 style attacks
-func newTLSConfig() *tls.Config {
-	return &tls.Config{
-		// MySQL Community edition has some problems with TLS1.2
-		// TODO: Validate this will not break servers using mysql community edition < 5.7.10
-		// MinVersion: tls.VersionTLS12,
-
-		// Default ordering taken from
-		// go 1.11 crypto/tls/cipher_suites.go
-		CipherSuites: []uint16{
-			tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
-			tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
-			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
-			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
-			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+func newTLSConfig(minVersion uint16) *tls.Config {
+
+	ciphers := []uint16{
+		tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+		tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+		tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+		tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+		tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
+		tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
+	}
+
+	if minVersion < tls.VersionTLS12 {
+		ciphers = append(ciphers, []uint16{
 			tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
 			tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
 			tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
 			tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
 			tls.TLS_RSA_WITH_AES_128_CBC_SHA,
 			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
-		},
+		}...)
+	}
+
+	return &tls.Config{
+		MinVersion:   minVersion,
+		CipherSuites: ciphers,
+	}
+}
+
+// TLSVersionToNumber converts a text description of the TLS protocol
+// to the internal Go number representation.
+func TLSVersionToNumber(tlsVersion string) uint16 {
+	switch tlsVersion {
+	case "TLSv1.3":
+		return tls.VersionTLS13
+	case "", "TLSv1.2":
+		return tls.VersionTLS12
+	case "TLSv1.1":
+		return tls.VersionTLS11
+	case "TLSv1.0":
+		return tls.VersionTLS10
+	default:
+		return tls.VersionTLS12
```

Comment:
Maybe a more sensible option here would be to return an error when we don't recognize a TLS string, instead of defaulting to 1.2. It's also better for future proofing!

Reply:
Done!